ADR/0028-component-dependencies.md

```diff
@@ -0,0 +1,179 @@
+* We will call a Component on which other Components have a dependency: a **"dependency"** component.
+* [integration-service] will handle the testing and promotion of all **dependancy** components in a special way.
+* [build-service] will use the `depends on` links to propose updates to users' **dependant** git repositories.
+* The set of Components and their "depends on" links form a directed-acyclic-graph (DAG). Application-service is responsible for making sure that cycles are not introduced and for managing references in the lifecycle of all Components in the Application. Deleting a Component should be well-defined.
+* As with Components in general, the "depends on" references are Application-scoped. Declared relationships between Components in different Applications are not to be considered valid. Declared relationships between Components in different namespaces are not to be considered valid.
+
+**PR Groups**: Introduce a new convention for PRs that lets one PR declare that it is related to another PR.
+
+* Treat the **source branch name** of the PR as a slash-delimited string, and use the first element of that
+  list as the PR Group name.
+* A new set of functionality for [integration-service] uses PR Groups to guide testing of related
+  PRs.
+* The set of PRs in a PR Group form a flat set.
+* PR Groups are scoped to Component repos under a single Application. Even if PRs across different
+  applications share the same **source branch name**, they shouldn't be considered part of the
+  same PR Group.
+* For example, if the source branch name for a PR is `my-feature/change-1`, then the PR Group
+  name should be interpreted as `my-feature`.
+* A source branch name with no slashes is the degenerate case. For example, if the source branch
+  name for a PR is `my-feature`, then the PR Group name should be interpreted as `my-feature`.
+* Since all PRs have a source branch name, all PRs are trivially members of a PR Group.
+
+Important to understand: the declared dependencies as defined on the Components' `depends on` lists and PR Groups are two different things:
+
+* Component dependencies are declarations about which Components' builds need each other. Automation uses this to create PRs in PR Groups.
+  * You say in english: “Builds of Component A depend on builds of Component B.”
+* PR Groups can be supplied by the user without any declared Component dependencies:
+  * You say in english: “PR #1 and PR #2 are in the same PR Group.”
+
+**Context directories** for [Pipelines as Code](https://pipelinesascode.com/)(PaC): introduce a new `spec.context` field in the `Repository` CR. When present, this field instructs PaC to only create a PipelineRun for events which include changes to the declared context directory in the git repository. This enables us to only run PipelineRuns that are relevant to Components that change in a given push or pull request event. By default, this is ".": the entire git repository.
```

Comment:
How will this work for monorepos where there are potentially different contexts?

Also what happens in the case where both the OLM bundle and the controller are built from the context root of the repository (e.g. in the gatekeeper operator lined above)?

Reply:
What are you asking here specifically with the first question? Is the other thread that Gal opened the same question effectively?

I think that this approach would run into an issue if the context directories are the same. Both Components would be rebuilding with every commit and therefore we would not be able to achieve consistency in the dependencies within the Snapshot.

Reply:
it's already supported by paac:

https://nightly.pipelines-as-code.pages.dev/docs/guide/authoringprs/#advanced-event-matching

(.pathChanged())

Reply:
(we can add another annotation to make it "easier" tho instead of having to use a CEL expression which may not be super friendly, but more flexible)

Reply:
I agree that there are challenges with depending on build contexts to separate the builds (and triggers). I think that the [other thread](https://github.com/redhat-appstudio/book/pull/113#discussion_r1301008578) sufficiently covers the concern presented in this one. To avoid confusion, I will close this one and leave the other one open. Please re-open this if you think that my judgement was inaccurate.